tidy3d/components/source.py

```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
         return const * offset * oscillation * amp
 
 
-SourceTimeType = Union[GaussianPulse, ContinuousWave]
+class CustomSourceTime(Pulse):
+    """Custom source time dependence.
+
+    Note
+    ----
+    To interpolate to simulation time steps, the value provided at the nearest time is used.
+
+    Example
+    -------
+    >>> sourcetime = CustomSourceTime.from_numpy(values=np.linspace(0, 1, 100), dt=.1)
+    """
+
+    sourcetime_dataset: Optional[TimeDataset] = pydantic.Field(
+        ..., title="Source time dataset", description="Dataset for storing the custom source time."
+    )
+
+    _sourcetime_dataset_none_warning = warn_if_dataset_none("sourcetime_dataset")
+
+    @classmethod
+    def from_numpy(
+        cls, freq0: float, fwidth: float, values: ArrayComplex1D, dt: float
+    ) -> "CustomSourceTime":
+        """Create a CustomSourceTime from a numpy array.
+
+        Parameters
+        ----------
+        freq0 : float
+            Estimated central frequency of the source.
+        fwidth : float
+            Estimated frequency width of the source.
+        values: ArrayComplex1D
+            Complex values of the source amplitude.
+        dt: float
+            Time step for the `values` array.
+
+        Returns
+        -------
+        CustomSourceTime
+            `CustomSourceTime` with these values.
+        """
+        times = np.arange(len(values)) * dt
+        return CustomSourceTime(
+            freq0=freq0,
+            fwidth=fwidth,
+            sourcetime_dataset=TimeDataset(values=TimeDataArray(values, coords=dict(t=times))),
```

Comment:
can we construct this variable before passing it to `CustomSourceTime` for readability?

```py
sourcetime_dataset=TimeDataArray(values, coords=dict(t=times))

return CustomSourceTime(
         freq0=freq0,
         fwidth=fwidth,
         sourcetime_dataset=sourcetime_dataset,
)